trie/triedb/pathdb/journal.go

```diff
@@ -0,0 +1,322 @@
+func (db *Database) loadDiffLayer(parent layer, r *rlp.Stream) (layer, error) {
+	// Read the next diff journal entry
+	var root common.Hash
+	if err := r.Decode(&root); err != nil {
+		// The first read may fail with EOF, marking the end of the journal
+		if err == io.EOF {
+			return parent, nil
+		}
+		return nil, fmt.Errorf("load diff root: %v", err)
+	}
+	var block uint64
+	if err := r.Decode(&block); err != nil {
+		return nil, fmt.Errorf("load block number: %v", err)
+	}
+	// Read in-memory trie nodes from journal
+	var encoded []journalNodes
+	if err := r.Decode(&encoded); err != nil {
+		return nil, fmt.Errorf("load diff nodes: %v", err)
+	}
+	nodes := make(map[common.Hash]map[string]*trienode.Node)
+	for _, entry := range encoded {
+		subset := make(map[string]*trienode.Node)
+		for _, n := range entry.Nodes {
+			if len(n.Blob) > 0 {
+				subset[string(n.Path)] = trienode.New(crypto.Keccak256Hash(n.Blob), n.Blob)
+			} else {
+				subset[string(n.Path)] = trienode.NewDeleted()
+			}
+		}
+		nodes[entry.Owner] = subset
+	}
+	// Read state changes from journal
+	var (
+		states     []journalStates
+		accounts   = make(map[common.Hash][]byte)
+		storages   = make(map[common.Hash]map[common.Hash][]byte)
+		incomplete = make(map[common.Hash]struct{})
+	)
+	if err := r.Decode(&states); err != nil {
+		return nil, fmt.Errorf("load diff states: %v", err)
+	}
+	for _, entry := range states {
+		set := make(map[common.Hash][]byte)
+		for _, n := range entry.States {
+			if len(n.Value) > 0 {
+				set[n.Key] = n.Value
+			} else {
+				set[n.Key] = nil
+			}
+		}
+		if entry.Owner == (common.Hash{}) {
+			accounts = set
+		} else {
+			storages[entry.Owner] = set
+		}
+		if entry.Incomplete {
+			incomplete[entry.Owner] = struct{}{}
+		}
+	}
+	return db.loadDiffLayer(newDiffLayer(parent, root, parent.stateID()+1, block, nodes, triestate.New(accounts, storages, incomplete)), r)
+}
+
+// journal implements the layer interface, marshaling the un-flushed trie nodes
+// along with layer meta data into provided byte buffer.
+func (dl *diskLayer) journal(w io.Writer) error {
+	// Ensure the layer didn't get stale
+	if dl.isStale() {
+		return errSnapshotStale
+	}
+	// Step one, write the disk root into the journal.
+	if err := rlp.Encode(w, dl.root); err != nil {
+		return err
+	}
+	// Step two, write the corresponding state id into the journal
+	if err := rlp.Encode(w, dl.id); err != nil {
+		return err
+	}
+	// Step three, write all unwritten nodes into the journal
+	nodes := make([]journalNodes, 0, len(dl.buffer.nodes))
+	for owner, subset := range dl.buffer.nodes {
+		entry := journalNodes{Owner: owner}
+		for path, node := range subset {
+			entry.Nodes = append(entry.Nodes, journalNode{Path: []byte(path), Blob: node.Blob})
+		}
+		nodes = append(nodes, entry)
+	}
+	if err := rlp.Encode(w, nodes); err != nil {
+		return err
+	}
+	log.Debug("Journaled disk layer", "root", dl.root, "nodes", len(dl.buffer.nodes))
```

Comment:
Nitpick, "Journaled pathdb disk layer", will help figuring out where this comes from.